File: Oct1/Lecture11Examples/WeatherRepository.java
package com.example.example35;

import android.app.Application;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;

import androidx.core.os.HandlerCompat;
import androidx.lifecycle.MutableLiveData;

import org.json.JSONException;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.net.URL;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class WeatherRepository {
    private static WeatherRepository instance;
    private final MutableLiveData<WeatherData> jsonData =
            new MutableLiveData<WeatherData>();
    private String mLocation;
    private String mJsonString;
    private WeatherDao mWeatherDao;



    private WeatherRepository(Application application){
        WeatherRoomDatabase db = WeatherRoomDatabase.getDatabase(application);
        mWeatherDao = db.weatherDao();
        if(mLocation!=null)
            loadData();
    }
    public static synchronized WeatherRepository getInstance(Application application){
        if(instance==null){
            instance = new WeatherRepository(application);
        }
        return instance;
    }

    public void setLocation(String location){
        mLocation = location;
        loadData();
        insert();
    }

    private void insert(){
        if(mLocation!=null && mJsonString!=null) {
            WeatherTable weatherTable = new WeatherTableBuilder().setLocation(mLocation).setWeatherJson(mJsonString).createWeatherTable();
            WeatherRoomDatabase.databaseExecutor.execute(() -> {
                mWeatherDao.insert(weatherTable);
            });
        }
    }

    public MutableLiveData<WeatherData> getData() {
        return jsonData;
    }

    private void loadData()
    {
        new FetchWeatherTask().execute(mLocation);
    }

    private class FetchWeatherTask{
        ExecutorService executorService = Executors.newSingleThreadExecutor();
        Handler mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper());
        public void execute(String location)
        {
            executorService.execute(() -> {
                String jsonWeatherData;
                URL weatherDataURL = NetworkUtils.buildURLFromString(location);
                jsonWeatherData = null;
                try{
                    jsonWeatherData = NetworkUtils.getDataFromURL(weatherDataURL);
                    if(jsonWeatherData!=null) {
                        mJsonString = jsonWeatherData;
                        postToMainThread(jsonWeatherData);
                    }
                }catch(Exception e){
                    e.printStackTrace();
                }
            });
        }

        private void postToMainThread(String retrievedJsonData){
            mainThreadHandler.post(() -> {
                try {
                    jsonData.setValue(JSONWeatherUtils.getWeatherData(retrievedJsonData));
                }catch(JSONException e){
                    e.printStackTrace();
                }
            });
        }
    }
}

File: Sep24/Example34/app/src/main/java/com/varunshankar/example34/WeatherRepository.kt
package com.varunshankar.example34

import com.varunshankar.example34.NetworkUtils.buildURLFromString
import com.varunshankar.example34.NetworkUtils.getDataFromURL
import com.varunshankar.example34.JSONWeatherUtils.getWeatherData
import android.app.Application
import androidx.lifecycle.MutableLiveData
import androidx.core.os.HandlerCompat
import android.os.Looper
import org.json.JSONException
import java.lang.Exception
import java.util.concurrent.Executors
import kotlin.jvm.Synchronized

class WeatherRepository private constructor(application: Application) {
    val data = MutableLiveData<WeatherData>()
    private var mLocation: String? = null
    fun setLocation(location: String?) {
        mLocation = location
        loadData()
    }

    private fun loadData() {
        FetchWeatherTask().execute(mLocation)
    }

    private inner class FetchWeatherTask {
        var executorService = Executors.newSingleThreadExecutor()
        var mainThreadHandler = HandlerCompat.createAsync(Looper.getMainLooper())
        fun execute(location: String?) {
            executorService.execute {
                var jsonWeatherData: String?
                val weatherDataURL = buildURLFromString(location!!)
                jsonWeatherData = null
                try {
                    jsonWeatherData = getDataFromURL(weatherDataURL!!)
                    jsonWeatherData?.let { postToMainThread(it) }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        private fun postToMainThread(retrievedJsonData: String) {
            mainThreadHandler.post {
                try {
                    data.setValue(getWeatherData(retrievedJsonData))
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            }
        }
    }

    companion object {
        @Volatile
        private var instance: WeatherRepository? = null
        @Synchronized
        fun getInstance(application: Application): WeatherRepository {
            if (instance == null) {
                instance = WeatherRepository(application)
            }
            return instance as WeatherRepository
        }
    }

    init {
        if (mLocation != null) loadData()
    }
}